### PR TITLE
[move-compiler] Make addresses in error messages more readable 

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

- About two years ago, we made named addresses transparent, instead of opaque, with respect to their numerical value. 
- To help prevent confusion, the compiler always expanded the named address. 
- Now, the compiler will only do so when the named address is not in a 1:1 mapping with its value 

## Test Plan 

- Added tests, updated tests 

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [X] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### Release notes

The Move compiler will now try to make addresses in errors more readable, expanding them to their numerical value only when it is potentially ambiguous.

### CODE REVIEW

File: external-crates/move/move-compiler/src/expansion/translate.rs [hunk 1]
```diff
@@ -122,6 +126,40 @@ impl<'env, 'map> Context<'env, 'map> {
...
+    let mut name_to_addr: BTreeMap<Symbol, BTreeSet<AccountAddress>> = BTreeMap::new();
+    let mut addr_to_name: BTreeMap<AccountAddress, BTreeSet<Symbol>> = BTreeMap::new();
+    let all_addrs = prog.named_address_maps.all().iter().chain(
+        pre_compiled_lib
+            .iter()
+            .flat_map(|pre| pre.parser.named_address_maps.all()),
+    );
+    for map in all_addrs {
+        for (n, addr) in map {
+            let n = *n;
```
Comment (tzakian) [inline external-crates/move/move-compiler/src/expansion/translate.rs:144]: Don't think this binding/dereference is needed? Unless this is mutating `n` in some way but as far as I can tell that isn't happening here.

File: external-crates/move/move-compiler/src/expansion/translate.rs [hunk 2]
```diff
@@ -122,6 +126,40 @@ impl<'env, 'map> Context<'env, 'map> {
...
+            let addr = addr.into_inner();
+            name_to_addr.entry(n).or_default().insert(addr);
+            addr_to_name.entry(addr).or_default().insert(n);
+        }
+    }
+    let name_to_addr_conflicts = name_to_addr
+        .into_iter()
+        .filter(|(_, addrs)| addrs.len() > 1)
+        .map(|(n, _)| n);
```
Comment (tzakian) [inline external-crates/move/move-compiler/src/expansion/translate.rs:153]: Nit: consider a filter_map here and below instead
```suggestion
        .filter_map(|(n, addrs)| if addrs.len() > 1 { Some(n) } else { None });
```

**Review by tzakian — APPROVED** [review]
Overall looks good to me! Two small nits.

File: external-crates/move/move-compiler/src/expansion/translate.rs (hunk 1 above)
Comment (tnowacki) [reply external-crates/move/move-compiler/src/expansion/translate.rs:144]: I need the non-reference for the `entry`. I just preferred this over writing `*n` in each occurrence.

File: external-crates/move/move-compiler/src/expansion/translate.rs (hunk 2 above)
Comment (tnowacki) [reply external-crates/move/move-compiler/src/expansion/translate.rs:153]: I feel that is less readable than a single filter+map. I tend to only reach for filter_map if I'm already dealing with options, or have many intermediate steps.

**Pull request merged**